contributors/design-proposals/federation-clusterselector.md

```diff
@@ -0,0 +1,78 @@
+# ClusterSelector Federated Resource Placement
+
+This document proposes a design for label based control over placement of
+Federated resources.
+
+Tickets:
+
+- https://github.com/kubernetes/kubernetes/issues/29887
+
+Authors:
+
+- Dan Wilson (emaildanwilson@github.com).
+- Nikhil Jindal (nikhiljindal@github).
+
+##  Background
+
+End users will often need a simple way to target a subset of clusters for deployment of resources. In some cases this will be for a specific cluster in other cases it will be groups of clusters.
+A few examples...
+
+1. Deploy the foo service to all clusters in Europe
+1. Deploy the bar service to cluster test15
+1. Deploy the baz service to all prod clusters globally
+
+Currently, it's possible to control placement decision of Federated ReplicaSets
+using the `federation.kubernetes.io/replica-set-preferences` annotation. This provides functionaility to change the number of ReplicaSets created on each Federated Cluster, by setting the quantity for each Cluster by Cluster Name. Since cluster names are required, in situations where clusters are add/removed from Federation it would require the object definitions to change in order to maintain the same configuration. From the example above, if a new cluster is created in Europe and added to federation, then the replica-set-preferences would need to be updated to include the new cluster name.
+
+This proposal is to provide placement decision support for all object types using Labels on the Federated Clusters as opposed to cluster names. The matching language currently used for nodeAffinity placement decisions onto nodes can be leveraged. 
+
+Carrying forward the examples from above...
+
+1. "location=europe"
+1. "someLabel exists"
+1. "environment notin ["qa", "dev"]
+
+## Design
+
+The proposed design uses a ClusterSelector annotation that has a value that is parsed into a struct definition that follows the same design as the [NodeSelector type used w/ nodeAffinity](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L1798) and will also use the [Matches function](https://github.com/kubernetes/apimachinery/blob/master/pkg/labels/selector.go#L172) of the apimachinery project to determine if an object should be sent on to federated clusters or not.
+
+In situations where objects are not to be forwarded to federated clusters, instead a delete api call will be made using the object definition. If the object does not exist it will be ignored.
+
+The federation-controller will be used to implement this with shared logic stored as utility functions to reduce duplicated code where appropriate.
+
+### End User Functionality 
+The annotation `federation.alpha.kubernetes.io/cluster-selector` is used on kubernetes objects to specify additional placement decisions that should be made. The value of the annotation will be a json object of type ClusterSelector which is an array of type ClusterSelectorRequirement.
```

Comment:
1. To clarify, the annotation is not just on pods. It will be allowed on all resources for example: ConfigMaps.
2. The long term imo plan should be to add a field for this (similar to what we are doing for NodeAffinity in kube. It started as an annotation and is now being turned into an API field), but we havent discussed it much.